src/ann_solo/utils.py

```diff
@@ -1,104 +1,338 @@
+        An iterator of the SSMs with assigned scores and q-values.
     """
-    for _, _, q, ssm in pyteomics.auxiliary.qvalues(
-            ssms, key=operator.attrgetter('search_engine_score'), reverse=True,
-            is_decoy=operator.attrgetter('is_decoy'), remove_decoy=True,
-            formula=1, correction=0, full_output=True):
-        ssm.q = q
-        if q <= fdr:
-            yield ssm
+    logging.debug(
+        "Compute features for semi-supervised scoring from %d SSMs", len(ssms)
+    )
+    # Create a Dataset with SSM features.
+    features = pd.DataFrame(_compute_ssm_features(ssms))
+    features["group"] = _get_ssm_groups(ssms, min_group_size) if grouped else 0
+    dataset = mokapot.dataset.LinearPsmDataset(
+        features,
+        target_column="is_target",
+        spectrum_columns="index",
+        peptide_column="sequence",
+        group_column="group",
+    )
+    if grouped:
+        logging.debug(
+            "Partitioned %d SSMs into %d groups",
+            len(ssms),
+            features["group"].nunique(),
+        )
+    # Define the mokapot model.
+    #   - Choice between a random forest, linear SVM, or no semi-supervised
+    #     learning.
+    #   - Features are preprocessed by standardizing them, removing
+    #     zero-variance features, and removing highly correlated features.
+    #   - We perform minimal tuning of TODO hyperparameters.
+    if model is None:
+        logging.debug("Calculate q-values based on the cosine similarity")
+        # Calculate q-values based on the cosine similarity.
+        confidences = dataset.assign_confidence(features["cosine"], True)
+    else:
+        logging.debug(
+            "Train semi-supervised %s model and score SSMs", model.upper()
+        )
+        scaler = make_pipeline(
+            StandardScaler(),
+            VarianceThreshold(),
+            CorrelationThreshold(0.95),
+        )
+        if model == "svm":
+            clf = mokapot.model.PercolatorModel(scaler, train_fdr=fdr)
+        elif model == "rf":
+            clf = mokapot.Model(
+                GridSearchCV(
+                    RandomForestClassifier(random_state=1),
+                    param_grid={
+                        "max_depth": [3, 5, 7, 9, None],
+                        "class_weight": [
+                            {0: neg, 1: pos}
+                            for neg in (0.1, 1, 10)
+                            for pos in (0.1, 1, 10)
+                        ],
+                    },
+                    refit=False,
+                    cv=3,
+                    n_jobs=-1,
+                ),
+                scaler,
+                train_fdr=fdr,
+            )
         else:
-            break
+            raise ValueError(
+                "Unknown semi-supervised machine learning model given"
+            )
+        # Train the mokapot model and combine the SSMs for all groups.
+        confidences, _ = mokapot.brew(dataset, clf, fdr, folds=5)
```

Comment:
Let's stick with the default 3-fold cross-validation as also used by Percolator. This is accepted practice by the community, and I don't see a special reason that 5-fold CV would be better.

Reply:
This is a hyperparameter that I want to investigate once the HPC is available again. If I recall well, the initial paper of Percolator mentioned a 10-fold CV and stops earlier if it converges already before. The default in Mokpot is 3-folds. We can try 3 / 5 / 10 in the 3 query files I am currently using.

Reply:
The number of CV folds isn't usually a hyperparameter that necessarily needs to be tuned, instead it depends on your computational budget. There are a few papers that explain the Percolator CV procedure in more detail (e.g. [Granholm et al.](http://www.biomedcentral.com/1471-2105/13/S16/S3)). It usually doesn't really make a difference for performance, so I think here we can safely stick to the default.

Reply:
Thanks Wout for sharing the reference with me. Interesting paper. In that case, we stick to the default k-fold CV used by Mokapot.